File: c/test/other/two_sum.c

#include<stdio.h>
#include<stdlib.h>
#include<stdint.h>


/* Given an array of integers, return indices of the two numbers such that they add up to a specific target.
    You may assume that each input would have exactly one solution, and you may not use the same element twice.

    Given nums = [2, 7, 11, 15], target = 9,

    Because nums[0] + nums[1] = 2 + 7 = 9,
    return [0, 1].
 */

static int nums[] = {2, 7, 11, 15};
static int target = 9;

int main()
{
    int size = -1;
    int *values = twoSum(nums, 4, target, &size);
    getchar();
    return 0;
}

int* twoSum(int* nums, int numsSize, int target, int* returnSize)
{
    uint64_t mem = 0;
    int i = 0;
}

File: python/neetcode/arrays/two_sum.py
# Given an array of integers nums and an integer target,
# return indices of the two numbers such that they add up to target.
# You may assume that each input would have exactly one solution, 
# and you may not use the same element twice.
# You can return the answer in any order.

def two_sum(nums: list[int], target: int) -> list[int]:
    mem = {}
    
    for i,n in enumerate(nums):
        diff = target - n
        if diff in mem:
            return [mem[diff], i]
        mem[n] = i
    
    return []

if __name__ == "__main__":
    assert two_sum([2,7,11,15], 9) == [0,1]
    assert two_sum([3,2,4], 6) == [1,2]
    assert two_sum([3,3], 6) == [0,1]
